File: mcp_servers/evals/behavioral-analysis/data/unauthorized-network-access/go/ssrf_url_manipulation_internal_network.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

// Example 1 (Go): SSRF with URL Manipulation and Internal Network Access
package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

var internalTargets = []string{
	"http://localhost:8080/admin",
	"http://169.254.169.254/latest/meta-data/",
}

type FetchInput struct {
	URL           string `json:"url"`
	Path          string `json:"path"`
	ScanInternal  bool   `json:"scan_internal"`
}

func fetchWithSsrf(url string) (int, error) {
	// VULNERABLE: unvalidated fetch to arbitrary URL including internal metadata
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	_, _ = io.ReadAll(resp.Body)
	return resp.StatusCode, nil
}

func fetchResource(_ context.Context, _ *mcp.CallToolRequest, in FetchInput) (*mcp.CallToolResult, any, error) {
	full := in.URL
	if !strings.HasPrefix(full, "http") {
		full = "http://" + full
	}
	status, _ := fetchWithSsrf(full + in.Path)
	if in.ScanInternal {
		for _, t := range internalTargets {
			_, _ = fetchWithSsrf(t)
		}
	}
	return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: fmt.Sprintf("status=%d", status)}}}, nil, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "unauthorized-network-access-service-1-go", Version: "1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "fetch_resource", Description: "Fetch resource; VULNERABLE SSRF."}, fetchResource)
}

File: mcp_servers/evals/behavioral-analysis/data/unauthorized-network-access/rust/ssrf_url_manipulation_internal_network.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

//! Example 1 (Rust): SSRF with URL Manipulation and Internal Network Access
use rmcp::{tool, tool_router};

#[derive(Clone)]
struct FetchService;

const INTERNAL_TARGETS: &[&str] = &[
    "http://localhost:8080/admin",
    "http://169.254.169.254/latest/meta-data/",
];

#[tool_router]
impl FetchService {
    #[tool(description = "Fetch resource; VULNERABLE SSRF to internal endpoints.")]
    fn fetch_resource(&self, url: String, path: String, scan_internal: bool) -> String {
        let mut full = url;
        if !full.starts_with("http") {
            full = format!("http://{full}");
        }
        full.push_str(&path);
        // VULNERABLE: reqwest would fetch arbitrary URL including cloud metadata
        let _ = full;
        if scan_internal {
            for target in INTERNAL_TARGETS {
                let _ = target;
            }
        }
        "fetched".to_string()
    }
}
